### ISSUE TITLE
Remove defined attribute from being set during element upgrading

### ISSUE THREAD

**Issue by treshugart** [skatejs/skatejs#1065]
Since we're close to shadow DOM support, there's the Shady CSS polyfill and the consumer can do the same thing, I think we should remove support for setting the `defined` attribute.

**Comment by Hotell** [comment]
hmm, ShadyCss does that?

**Comment by Hotell** [comment]
also [ShadyCss doesn't work with skate](https://github.com/skatejs/skatejs/issues/1027) at this moment, and probably won't in the future, so the question is why use it at all.

About the define, we are adding it just to "polyfill" `:defined` in non native browsers right?

**Comment by treshugart** [comment]
I believe they polyfill `:defined`, but I could be wrong. I actually got it working before, but ever since they started removing the style tag, I couldn't get the transformed css from it. There must be a way; it feels far to connected to a single particular way to write web components using <template> which isn't ideal for many.

**Comment by Hotell** [comment]
So they are not polyfiling it http://codepen.io/Hotell/pen/ggQELy?editors=1010 

AFAIR it is not possible, to polyfill it

**Comment by Hotell** [comment]
<img width="1015" alt="screen shot 2017-02-11 at 2 08 58 pm" src="https://cloud.githubusercontent.com/assets/1223799/22853996/cc2e0f46-f063-11e6-90bc-069ca2b56835.png">

**Comment by Hotell** [comment]
https://developers.google.com/web/fundamentals/getting-started/primers/customelements#polyfill

TL;DR

>  Note: the :defined CSS pseudo-class cannot be polyfilled.

**Comment by thepian** [comment]
I think it would be nice to know what the target for Skate is conceptionally in terms of browser support and the type of website. It currently seems aimed at supporting prototyping rather than mainstream production, but that would of course shift as older browsers fall off the grid. 

If I understand this ticket correctly, it implies that you will aggressively move forward, which sounds correct for prototypes, but not good for building something widely deployed.

**Comment by treshugart** [comment]
Interesting. It'd be possible if the ShadyCSS polyfill hooked into the custom Custom Element polyfill. `<custom-style>` would be able to transform `:defined` to something that the CE fill adds when a component is upgraded. Why wouldn't that be polyfillable?

Maybe it's worth leaving this in then. I'm wary of the performance implications as attribute sets are three times slower than property sets and doing that for every upgrade across the board seems a bit much when the consumer can opt into it by simply doing it manually, using decorators or using mixins.

**Comment by thepian** [comment]
I think turning Skate into a bag of tricks(mixins) to hook onto a very thin skeleton is a very good idea. 

```
@skate({ props, renderer, postUpgrade})
class Foo extends HTMLElement {
}
```

would allow for trying out competing solutions and varying strategy by component.

**Comment by treshugart** [comment]
That's kinda the direction it's going, except that it doesn't use the decorator API. The class API already allows you to integrate with other solutions and allows control over the order of composition, if necessary. Additionally, it's stable API now, whereas the decorator API is still in stage 2, so probably better to base the most things on the classes.

**Comment by treshugart** [comment]
This was removed, but is easy for consumers to implement. See PR referenced above for details.
